src/probability/variance.lean

```diff
@@ -30,28 +39,166 @@ open_locale big_operators measure_theory probability_theory ennreal nnreal
 
 namespace probability_theory
 
-/-- The variance of a random variable is `𝔼[X^2] - 𝔼[X]^2` or, equivalently, `𝔼[(X - 𝔼[X])^2]`. We
-use the latter as the definition, to ensure better behavior even in garbage situations. -/
+/-- The `ℝ≥0∞`-valued variance of a real-valued random variable defined as a lintegral.-/
+def evariance {Ω : Type*} {m : measurable_space Ω} (f : Ω → ℝ) (μ : measure Ω) : ℝ≥0∞ :=
+∫⁻ ω, ∥f ω - μ[f]∥₊^2 ∂μ
+
+/-- The `ℝ`-valued variance of a real-valued random variable defined by applying `ennreal.to_real`
+to `evariance`. -/
 def variance {Ω : Type*} {m : measurable_space Ω} (f : Ω → ℝ) (μ : measure Ω) : ℝ :=
-μ[(f - (λ ω, μ[f])) ^ 2]
+(evariance f μ).to_real
+
+lemma _root_.measure_theory.mem_ℒp.evariance_lt_top {Ω : Type*} {m : measurable_space Ω}
+  {f : Ω → ℝ} {μ : measure Ω} [is_finite_measure μ] (hf : mem_ℒp f 2 μ) :
+  evariance f μ < ∞ :=
+begin
+  have := ennreal.pow_lt_top (hf.sub $ mem_ℒp_const $ μ[f]).2 2,
+  rw [snorm_eq_lintegral_rpow_nnnorm ennreal.two_ne_zero ennreal.two_ne_top,
+    ← ennreal.rpow_two] at this,
+  simp only [pi.sub_apply, ennreal.to_real_bit0, ennreal.one_to_real, one_div] at this,
+  rw [← ennreal.rpow_mul, inv_mul_cancel (two_ne_zero : (2 : ℝ) ≠ 0), ennreal.rpow_one] at this,
+  simp_rw ennreal.rpow_two at this,
+  exact this,
+end
+
+lemma evariance_eq_top {Ω : Type*} {m : measurable_space Ω} {f : Ω → ℝ}
+  {μ : measure Ω} [is_finite_measure μ] (hfm : ae_strongly_measurable f μ) (hf : ¬ mem_ℒp f 2 μ) :
+  evariance f μ = ∞ :=
+begin
+  by_contra h,
+  rw [← ne.def, ← lt_top_iff_ne_top] at h,
+  have : mem_ℒp (λ ω, f ω - μ[f]) 2 μ,
+  { refine ⟨hfm.sub ae_strongly_measurable_const, _⟩,
+    rw snorm_eq_lintegral_rpow_nnnorm ennreal.two_ne_zero ennreal.two_ne_top,
+    simp only [ennreal.to_real_bit0, ennreal.one_to_real, ennreal.rpow_two, ne.def],
+    exact ennreal.rpow_lt_top_of_nonneg (by simp) h.ne },
+  refine hf _,
+  convert this.add (mem_ℒp_const $ μ[f]),
+  ext ω,
+  rw [pi.add_apply, sub_add_cancel],
+end
+
+lemma evariance_lt_top_iff_mem_ℒp {Ω : Type*} {m : measurable_space Ω} {f : Ω → ℝ}
+  {μ : measure Ω} [is_finite_measure μ] (hf : ae_strongly_measurable f μ) :
+  evariance f μ < ∞ ↔ mem_ℒp f 2 μ :=
+begin
+  refine ⟨_, measure_theory.mem_ℒp.evariance_lt_top⟩,
+  contrapose,
+  rw [not_lt, top_le_iff],
+  exact evariance_eq_top hf
+end
+
+lemma _root_.measure_theory.mem_ℒp.of_real_variance_eq {Ω : Type*} {m : measurable_space Ω}
+  {f : Ω → ℝ} {μ : measure Ω} [is_finite_measure μ] (hf : mem_ℒp f 2 μ) :
+  ennreal.of_real (variance f μ) = evariance f μ :=
+begin
+  rw [variance, ennreal.of_real_to_real],
+  exact hf.evariance_lt_top.ne,
+end
+
+lemma evariance_eq_lintegral_of_real {Ω : Type*} {m : measurable_space Ω}
+  (f : Ω → ℝ) (μ : measure Ω) :
+  evariance f μ = ∫⁻ ω, ennreal.of_real ((f ω - μ[f])^2) ∂μ :=
+begin
+  rw evariance,
+  congr,
+  ext1 ω,
```

Comment:
```suggestion
  congr' with ω,
```
(untested, and sometimes congr' does not give the same thing as congr)

Reply:
It doesn't work here since `ext` goes too far. I was able to use this trick elsewhere in the file however. Thanks for letting me know!